src/NuGet.Core/NuGet.Build.Tasks/NuGet.props

```diff
@@ -38,4 +38,7 @@ Copyright (c) .NET Foundation. All rights reserved.
     <CentralPackageVersionsFileImported>true</CentralPackageVersionsFileImported>
   </PropertyGroup>
 
+  <!-- Extension point after Directory.Packages.props is imported. -->
+  <Import Project="$(CustomAfterDirectoryPackagesProps)" Condition="'$(CentralPackageVersionsFileImported)' == 'true' and '$(CustomAfterDirectoryPackagesProps)' != ''" />
```

Comment:
I feel like building .NET projects is already too slow, and adding in extra "check if this file name exists in every single parent directory up to the root" is not helping performance. Sure, the operating system will (hopefully) cache directory listings, but even checking cached data takes more time than not checking.

There's already a Directory.Build.targets file that will get imported after Directory.Packages.props file. Yes, it will get imported much later. Even the contents of the csproj will be evaluated after the Directory.Packages.props file.

So, I don't understand the benefit or use-case, and I'm wondering if we as .NET more broadly should think more about build performance?

Reply:
Maybe I care about this more than I should?

<details>
<div>

```cs
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

public class Program
{
    private static void Main(string[] args)
    {
        BenchmarkRunner.Run<Program>();
    }

    [Benchmark]
    public string? FindFile()
    {
        string? directory = Directory.GetCurrentDirectory();
        while (!string.IsNullOrEmpty(directory))
        {
            string file = Path.Combine(directory, "Directory.Whatever.props");
            if (File.Exists(file))
            {
                return file;
            }

            directory = Path.GetDirectoryName(directory);
        }

        return null;
    }
}
```

</div>
</details>

| Method   | Mean     | Error   | StdDev  | Gen0   | Allocated |
|--------- |---------:|--------:|--------:|-------:|----------:|
| FindFile | 124.4 us | 0.70 us | 0.66 us | 0.2441 |   4.56 KB |

This is just one file, so multiple this by every file that the .NET build search for (at least 3, but maybe more), multiplied by the number of projects in the solution. For huge repos with 1000+ projects, that's going to be hundreds of milliseconds (maybe even over a second?) doing nothing more than searching for files.

This is also an extensibility point to load a customer provided file after loading another customer provided file.  Why can't the customer just add the `<Import` at the end of their own Directory.Packages.props?

Reply:
I just noticed this PR is NOT using `[MSBuild]::GetDirectoryNameOfFileAbove`, so it's not anywhere near as bad as my initial gut reaction.

I still don't understand why any of these before/after imports extensibility points exist. External customers can put the `<Import` in their own `Directory.*.[props|targets]` files.  If this is only needed for the .NET virtual mono repo, then it seems more appropriate to me to enforce all products that the VRM consists of to add the imports, rather than change the product which affects all customers. 🤷

Reply:
This feature request isn't tied to the VMR effort. We want to make use of this in individual repo builds as well. Note that there's also a customer asking for this.

I took a look and couldn't find an extension point that allows to import logic after D.P.props gets imported and before the project file gets read. That's the hook point that we need.

Perf isn't a reason to reject this. It is completely opt-in and we already have 100+ imports in a normal .NET project.